### PR TITLE
Fix pack ordering for `PackType.SERVER_DATA`

### PR DESCRIPTION
**Pull request by Technici4n**
The first fix in this PR is to preserve the pack ordering in `PackResources#discoverAvailable`. Vanilla orders packs by name there, which doesn't really make sense for ordering purposes across sources. Instead, we use a linked hash map to preserve the ordering between the `RepositorySource`s, but continue to sort by name the packs within a source.

The second fix is to always fire `AddPackFindersEvent` **after** the `RepositorySource` for mod builtin packs is added, with the introduction of the new `populatePackRepository` method that always handles the two in the right order. This ensures that repository sources from `AddPackFindersEvent` can properly override mod builtin packs.

The data pack order is stored in the world data, hence you will need to re-create worlds to observe the effects of this PR.

### CODE REVIEW

File: patches/net/minecraft/server/packs/repository/PackRepository.java.patch [hunk 1]
```diff
@@ -9,6 +9,15 @@
      }
  
      public void reload() {
+@@ -33,7 +_,7 @@
+     }
+ 
+     private Map<String, Pack> discoverAvailable() {
+-        Map<String, Pack> map = Maps.newTreeMap();
++        Map<String, Pack> map = Maps.newLinkedHashMap(); // Neo: use a linked hash map to preserve ordering
```
Comment (embeddedt) [inline patches/net/minecraft/server/packs/repository/PackRepository.java.patch]: IMO we still need to sort the packs produced by each `RepositorySource` before adding them to the map. The previous code would be able to handle sources that produce packs in a non-deterministic order; this implementation cannot as it only sorts by order of sources, not the packs they produce.

However, someone more familiar with the pack handling code can correct me if the order doesn't matter.

Comment (Technici4n) [reply patches/net/minecraft/server/packs/repository/PackRepository.java.patch]: The entire point is to preserve ordering - blindly sorting won't really help. I can check vanilla's sources, but the output of modded sources should not be sorted IMO.

Comment (Technici4n) [reply patches/net/minecraft/server/packs/repository/PackRepository.java.patch]: OK you were right. Now the sorting only happens within each source. That should be acceptable (and still much better than the current behavior).

**Review by embeddedt — APPROVED** [review]
New strategy is exactly what I was thinking

**Pull request merged**